radix-engine/src/transaction/transaction_executor.rs

```diff
@@ -726,38 +720,43 @@ where
         (fee_summary, fee_payments)
     }
 
-    fn update_intent_hash_store(
+    fn update_transaction_tracker(
         track: &mut Track<S, SpreadPrefixKeyMapper>,
         next_epoch: Epoch,
-        intent_hash: Hash,
-        expiry_epoch: Option<Epoch>,
+        intent_hash: &TransactionIntentHash,
         is_success: bool,
     ) {
         // Read the intent hash store
         let handle = track
             .acquire_lock(
-                INTENT_HASH_STORE.as_node_id(),
+                TRANSACTION_TRACKER.as_node_id(),
                 MAIN_BASE_PARTITION,
-                &IntentHashStoreField::IntentHashStore.into(),
+                &TransactionTrackerField::TransactionTracker.into(),
                 LockFlags::MUTABLE,
             )
             .unwrap()
             .0;
-        let mut substate: IntentHashStoreSubstate =
+        let mut transaction_tracker: TransactionTrackerSubstate =
             track.read_substate(handle).0.as_typed().unwrap();
 
         // Update the status of the intent hash
-        if let Some(expiry_epoch) = expiry_epoch {
-            if let Some(partition_number) = substate.partition_of(expiry_epoch.number()) {
+        if let TransactionIntentHash::ToCheck {
+            expiry_epoch,
+            intent_hash,
+        } = intent_hash
+        {
+            if let Some(partition_number) =
```

Comment:
If this isn't true, (eg due to epoch jumping forward, or going backward) we should panic IMO - because it's an invariant break? We should protect against replay not being tracked.

(And for this, I quite like the `let .. else .. ` construct)